Add algorithms for geometry tracking

sethrj opened the pull request:
Sorting, partitioning, and generic lower bound are required for some of the tracking algorithms. Since those use swap, move, and forward—none of which are device-compatible and also can't be overloaded due to namespace lookup ambiguity—I've added special device-friendly versions of them prefixed with a `c`.

pcanal commented on src/base/Algorithms.hh:
```diff
@@ -19,14 +20,14 @@ namespace celeritas
 template<class T>
 CELER_CONSTEXPR_FUNCTION T&&
-forward(typename std::remove_reference<T>::type& v) noexcept
+cforward(typename std::remove_reference<T>::type& v) noexcept
```
For the record, could you augment the git log to explain why/how this solution is superior to 'simply' requiring to always use the fully qualified name (`celeritas::forward`).

sethrj replied:
Will do. The problem is essentially that algorithms defined in the `std` namespace use unqualified names (`swap`) to enable argument-dependent lookup to find custom `swap` methods.

We might be able to get away with `forward` if we always qualify it with `celeritas::`, but I think I hit issues in testing when calling our algorithms on `std` containers/iterators.

sethrj replied:
Changed as noted below.

pcanal commented on src/base/Stopwatch.hh:
```diff
@@ -34,7 +34,7 @@ class Stopwatch
 
     // Get the current elapsed time [s]
-    inline real_type operator()() const;
+    inline double operator()() const;
```
Could enhance the git log to record the reason behind this change?

sethrj replied:
Yeah. Basically this is a CPU-only operation for testing/diagnostics, not related to any physics or on-device code, so we should use the "native" higher-precision type.

pcanal commented on src/base/detail/AlgorithmsImpl.hh:
```diff
@@ -0,0 +1,305 @@
+template<typename Integral>
+CELER_CONSTEXPR_FUNCTION
+    typename std::enable_if<std::is_integral<Integral>::value, Integral>::type
+    half_positive(Integral value)
```
Some algorithms start with `c` and some do not start with `c` with the discriminant being (if I understood correctly) "name is already use in `std` namespace".   [If the prefix stays necessary, see oher comment], should't we for clarity/regularity-of-pattern use the prefix for all algorithm (that have the `CELER_..._FUNCTION` decoration)?

sethrj replied:
That's a good point... right now the criteria is "do the std algorithms we use on celeritas data structures (or vice versa) use these functions/algorithms": and that's pretty arbitrary. I don't know offhand if the standard specifies whether *their* internal function usage has to be fully qualified, or which functions besides swap are supposed to be unqualified so they can use ADL.

amandalund reviewed (approved):
New algorithms look good @sethrj!

sethrj commented:
I forgot to annotate where the LLVM files came from, but for the record, here's the provenance and changes:
- Commit [b07b5bd72716625e0976a84d23652d94d8d0165a](https://github.com/llvm/llvm-project/commit/b07b5bd72716625e0976a84d23652d94d8d0165a)
- Inside `libcxx/include/__algorithm`: files `half_positive.h`, `lower_bound.h`, `partition.h`, `sift_down.h`, `pop_heap.h`, `make_heap.h`, `sort_heap.h`, `partial_sort.h`
- Cleaned variable/template names (removing leading underscores, tweaking other names)
- Formatted with our clang-format
- Added `difference_type_t`, `trivial_move`
- Replaced `swap` and `move` with `trivial_swap`, `trivial_move`

pcanal commented on src/base/Algorithms.hh:
```diff
@@ -43,6 +44,24 @@ CELER_CONSTEXPR_FUNCTION auto move(T&& v) noexcept ->
+                  "Value is not trivially movable");
+    static_assert(std::is_trivially_destructible<T>::value,
+                  "Value is not trivially destructible");
+    T temp{a};
```
Couldn't this be:
```
T temp{::celeritas::move(a)};
```
(and/or why does the other benefit from move and no that one) ?

sethrj replied:
Yes, you're completely right... not sure how I missed that. Probably in the shuffle of redefining swaps and moves and...
